### Task1


In [42]:
import pandas as pd
from statsmodels.stats.proportion import proportions_ztest
import numpy as np

In [43]:
df=pd.read_csv("ad_click_dataset.csv")
df

id full_name   age      gender device_type ad_position  \
0      670   User670  22.0         NaN     Desktop         Top   
1     3044  User3044   NaN        Male     Desktop         Top   
2     5912  User5912  41.0  Non-Binary         NaN        Side   
3     5418  User5418  34.0        Male         NaN         NaN   
4     9452  User9452  39.0  Non-Binary         NaN         NaN   
...    ...       ...   ...         ...         ...         ...   
9995  8510  User8510   NaN         NaN      Mobile         Top   
9996  7843  User7843   NaN      Female     Desktop      Bottom   
9997  3914  User3914   NaN        Male      Mobile        Side   
9998  7924  User7924   NaN         NaN     Desktop         NaN   
9999  3056  User3056  44.0        Male      Tablet         Top   

     browsing_history time_of_day  click  
0            Shopping   Afternoon      1  
1                 NaN         NaN      1  
2           Education       Night      1  
3       Entertainment     Evening      1  
4        Social Media     Morning      0  
...               ...         ...    ...  
9995        Education         NaN      0  
9996    Entertainment         NaN      0  
9997              NaN     Morning      0  
9998         Shopping     Morning      1  
9999     Social Media     Morning      0  

[10000 rows x 9 columns]

In [44]:
df = df.dropna(subset=['click'])
df = df.dropna(subset=['gender'])
df = df.dropna(subset=['ad_position'])


In [45]:
df

id full_name   age      gender device_type ad_position  \
1     3044  User3044   NaN        Male     Desktop         Top   
2     5912  User5912  41.0  Non-Binary         NaN        Side   
5     5942  User5942   NaN  Non-Binary         NaN      Bottom   
6     7808  User7808  26.0      Female     Desktop         Top   
7     5065  User5065  40.0        Male      Mobile        Side   
...    ...       ...   ...         ...         ...         ...   
9986  5574  User5574  52.0      Female     Desktop      Bottom   
9990  9540  User9540  64.0  Non-Binary         NaN        Side   
9996  7843  User7843   NaN      Female     Desktop      Bottom   
9997  3914  User3914   NaN        Male      Mobile        Side   
9999  3056  User3056  44.0        Male      Tablet         Top   

     browsing_history time_of_day  click  
1                 NaN         NaN      1  
2           Education       Night      1  
5        Social Media     Evening      1  
6                 NaN         NaN      1  
7                 NaN     Evening      0  
...               ...         ...    ...  
9986         Shopping   Afternoon      1  
9990        Education         NaN      0  
9996    Entertainment         NaN      0  
9997              NaN     Morning      0  
9999     Social Media     Morning      0  

[4221 rows x 9 columns]

In [46]:
print(df['gender'].unique())
print(df['ad_position'].unique())

['Male' 'Non-Binary' 'Female']
['Top' 'Side' 'Bottom']


In [47]:
df['gender'] = df['gender'].map({'Male': 0, 'Female': 1, 'Non-Binary': 2})
df['ad_position'] = df['ad_position'].map({'Top':0,'Bottom':1,'Side':2})


In [48]:
df

id full_name   age  gender device_type  ad_position browsing_history  \
1     3044  User3044   NaN       0     Desktop            0              NaN   
2     5912  User5912  41.0       2         NaN            2        Education   
5     5942  User5942   NaN       2         NaN            1     Social Media   
6     7808  User7808  26.0       1     Desktop            0              NaN   
7     5065  User5065  40.0       0      Mobile            2              NaN   
...    ...       ...   ...     ...         ...          ...              ...   
9986  5574  User5574  52.0       1     Desktop            1         Shopping   
9990  9540  User9540  64.0       2         NaN            2        Education   
9996  7843  User7843   NaN       1     Desktop            1    Entertainment   
9997  3914  User3914   NaN       0      Mobile            2              NaN   
9999  3056  User3056  44.0       0      Tablet            0     Social Media   

     time_of_day  click  
1            NaN      1  
2          Night      1  
5        Evening      1  
6            NaN      1  
7        Evening      0  
...          ...    ...  
9986   Afternoon      1  
9990         NaN      0  
9996         NaN      0  
9997     Morning      0  
9999     Morning      0  

[4221 rows x 9 columns]

In [49]:
groupa=df[df['ad_position']==0]

groupb=df[df['ad_position']==1]

In [50]:
groupa

id full_name   age  gender device_type  ad_position browsing_history  \
1     3044  User3044   NaN       0     Desktop            0              NaN   
6     7808  User7808  26.0       1     Desktop            0              NaN   
18    2124  User2124   NaN       0     Desktop            0              NaN   
28    6724  User6724  37.0       1         NaN            0    Entertainment   
31    1729  User1729  26.0       1      Mobile            0              NaN   
...    ...       ...   ...     ...         ...          ...              ...   
9960  9638  User9638  64.0       2     Desktop            0    Entertainment   
9964   373   User373  34.0       1      Mobile            0              NaN   
9981  7808  User7808  26.0       1     Desktop            0              NaN   
9982  3574  User3574   NaN       1      Tablet            0    Entertainment   
9999  3056  User3056  44.0       0      Tablet            0     Social Media   

     time_of_day  click  
1            NaN      1  
6            NaN      1  
18       Evening      1  
28       Evening      1  
31     Afternoon      1  
...          ...    ...  
9960     Morning      0  
9964     Evening      1  
9981       Night      1  
9982     Evening      1  
9999     Morning      0  

[1419 rows x 9 columns]

In [51]:
groupb

id full_name   age  gender device_type  ad_position browsing_history  \
5     5942  User5942   NaN       2         NaN            1     Social Media   
8     7993  User7993   NaN       2      Mobile            1     Social Media   
16    8127  User8127   NaN       0      Tablet            1              NaN   
17     188   User188  56.0       1      Tablet            1             News   
25    4890  User4890  43.0       0      Tablet            1        Education   
...    ...       ...   ...     ...         ...          ...              ...   
9975  9871  User9871  41.0       0     Desktop            1              NaN   
9979  8337  User8337   NaN       1         NaN            1    Entertainment   
9983  4254  User4254   NaN       0     Desktop            1              NaN   
9986  5574  User5574  52.0       1     Desktop            1         Shopping   
9996  7843  User7843   NaN       1     Desktop            1    Entertainment   

     time_of_day  click  
5        Evening      1  
8            NaN      1  
16           NaN      1  
17       Morning      1  
25     Afternoon      1  
...          ...    ...  
9975         NaN      0  
9979       Night      0  
9983   Afternoon      1  
9986   Afternoon      1  
9996         NaN      0  

[1513 rows x 9 columns]

In [52]:
a=groupa['click'].sum()

b=groupb['click'].sum()

lena=len(groupa)
lenb=len(groupb)

counting=[a,b]
lens=[lena,lenb]

zscore,pvalue=proportions_ztest(counting,lens)

print(f"Z-Score: {zscore}")
print(f"P-Value: {pvalue}")

Z-Score: -2.9832500834568774
P-Value: 0.002852048503930551


Yes, there is statistically significant difference in the click rates of top and bottom ad_position

* As z score is -2.98 the difference between 2 groups click rates is 3 standard deviations away from the mean under null hypothesis.
* p value is 0.00285 which is much less than 0.05, hence we reject null hypothesis which stated that there is no difference in click rates in the top and bottom ad-position
